mtmd: Add DeepSeekOCR Support

sfallah opened the pull request:
Feature Request: https://github.com/ggml-org/llama.cpp/issues/16676

*Make sure to read the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

### GGUF Models

https://huggingface.co/ggml-org/DeepSeek-OCR-GGUF/tree/main

### Running the Model

#### Build llama.cpp (Mac)
```
cmake -B build -G Ninja -DCMAKE_BUILD_TYPE=Release
cmake --build build -j --config Release
```

#### Running llama-server

```
./llama-mtmd-cli -hf ggml-org/DeepSeek-OCR-GGUF --top-k 1
```

Note: do not add `--chat-template deepseek-ocr` on server

Run chat completion with one single short prompt, example:
- `OCR`
- `OCR markdown`
- `<|grounding|>OCR`

Example:

```json
{
  "messages": [
    {
      "role": "user",
      "content": [
        {
          "type": "text",
          "text": "OCR"
        },
        {
          "type": "image_url",
          "image_url": {
            "url": "data:image/jpeg;base64,/9j/4AAQSk....."
          }
        }
      ]
    }
  ]
}
```

#### Running llama-mtmd-cli

##### DeepSeekOCR Paper (First page)

```
build/bin/llama-mtmd-cli \
-m gguf_models/deepseek-ai/deepseek-ocr-f16.gguf \
--mmproj gguf_models/deepseek-ai/mmproj-deepseek-ocr-f16.gguf \
--image tmp/mtmd_test_data/Deepseek-OCR-2510.18234v1_page1.png \
-p "<|grounding|>Convert the document to markdown." \
--chat-template deepseek-ocr --temp 0
```

##### Hard Test (Old Newspaper Image)

```
build/bin/llama-mtmd-cli \
-m gguf_models/deepseek-ai/deepseek-ocr-f16.gguf \
--mmproj gguf_models/deepseek-ai/mmproj-deepseek-ocr-f16.gguf \
--image tools/mtmd/test-1.jpeg \
-p "<|grounding|>Convert the document to markdown." \
--chat-template deepseek-ocr --temp 0
```

sfallah committed:
mtmd: llama.cpp DeepSeekOCR support
init commit

sfallah committed:
loading sam tensors

bluebread committed:
mtmd: fix vision model processing

sfallah committed:
Merge pull request #1 from bluebread/sf/deepseek-ocr

mtmd: fix vision model processing

sfallah committed:
deepseek-ocr clip-vit model impl

bluebread committed:
mtmd: add DeepSeek-OCR LM support with standard attention

bluebread committed:
mtmd: successfully runs DeepSeek-OCR LM in llama-cli

bluebread committed:
mtmd: Fix RoPE type for DeepSeek-OCR LM.

bluebread committed:
Merge branch 'sf/deepseek-ocr' of github.com:sfallah/llama.cpp into sf/deepseek-ocr

sfallah committed:
loading LM
testing Vision model loading

sfallah committed:
Merge branch 'sf/deepseek-ocr' into sf/deepseek-ocr

sfallah committed:
Merge pull request #2 from bluebread/sf/deepseek-ocr

mtmd: DeepseekOCR Implement DeepSeek3B-MoE-A570M (LM component)

sfallah committed:
sam warmup working

sfallah committed:
sam erroneous return corrected

sfallah committed:
clip-vit:  corrected cls_embd concat

sfallah committed:
clip-vit: model convert  qkv_proj split

sfallah committed:
corrected combining of image encoders' results

bluebread committed:
fix: update callback for ffn_moe_weighted and add callback for attn_out in deepseek2 model

bluebread committed:
Merge branch 'sf/deepseek-ocr' of github.com:sfallah/llama.cpp into sf/deepseek-ocr

sfallah committed:
concat image_newline and image_seperator tokens

sfallah committed:
visual_model warmup (technically) works

sfallah committed:
window partitioning using standard ggml ops

bluebread committed:
Merge branch 'sf/deepseek-ocr' of github.com:sfallah/llama.cpp into sf/deepseek-ocr

sfallah committed:
sam implementation without using CPU only ops

bluebread committed:
clip: fixed warnings

bluebread committed:
Merge branch 'sf/deepseek-ocr' of github.com:sfallah/llama.cpp into sf/deepseek-ocr

bluebread committed:
mtmd: fix get_rel_pos

bluebread committed:
Merge branch 'sf/deepseek-ocr' of github.com:sfallah/llama.cpp into sf/deepseek-ocr

bluebread committed:
mtmd: fixed the wrong scaler for get_rel_pos

sfallah committed:
image encoding technically works but the output can't be checked singe image decoding fails

bluebread committed:
mtmd: minor changed

bluebread committed:
Merge branch 'sf/deepseek-ocr' of github.com:sfallah/llama.cpp into sf/deepseek-ocr

sfallah committed:
Merge pull request #3 from bluebread/sf/deepseek-ocr

Fixed get_rel_pos & add_rel_pos_inplace operator

bluebread committed:
mtmd: add native resolution support

sfallah committed:
- image encoding debugged
- issues fixed mainly related wrong config like n_patches etc.
- configs need to be corrected in the converter

bluebread committed:
mtmd: correct token order

sfallah committed:
Merge pull request #5 from bluebread/dsocr-debug

debug: correct token order

sfallah committed:
Merge branch 'sf/deepseek-ocr' into sf/deepseek-ocr

sfallah committed:
Merge pull request #4 from bluebread/sf/deepseek-ocr

Add native resolution support

sfallah committed:
- dynamic resizing
- changes are concerning PR https://github.com/sfallah/llama.cpp/pull/4

bluebread committed:
mtmd: quick fix token order

bluebread committed:
mtmd: fix danling pointer

sfallah committed:
Merge pull request #6 from bluebread/sf/deepseek-ocr

mtmd: quick fix token order

bluebread committed:
mtmd: SAM numerically works

bluebread committed:
mtmd: debug CLIP-L (vit_pre_ln)

bluebread committed:
mtmd: debug CLIP-L & first working DeepSeek-OCR model

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	convert_hf_to_gguf.py
#	src/llama-model.cpp
#	src/models/deepseek2.cpp

bluebread committed:
Merge branch 'sf/deepseek-ocr' of github.com:sfallah/llama.cpp into sf/deepseek-ocr

bluebread committed:
mtmd : add --dsocr-mode CLI argument for DeepSeek-OCR resolution control & all native resolution modes work

bluebread committed:
mtmd: simplify SAM patch embedding

sfallah committed:
Merge pull request #7 from bluebread/sf/deepseek-ocr

First DeepSeek-OCR working implementation

sfallah committed:
Merge branch 'master' into sf/deepseek-ocr

# Conflicts:
#	convert_hf_to_gguf.py
#	tools/mtmd/clip.h
#	tools/mtmd/mtmd.cpp

bluebread commented on common/arg.cpp:
```diff
@@ -1829,6 +1829,21 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
+        "DeepSeek-OCR resolution mode, one of:\n"
+        "- auto (default): automatically select resolution\n"
+        "- tiny, small, base, large: native resolution\n"
+        "- gundam, gundam-master: dynamic resolution",
```
ok. I'll remove it later.

bluebread committed:
mtmd: adapt Pillow image resizing function

bluebread committed:
mtmd:  simplify DeepSeek-OCR dynamic resolution preprocessing

bluebread committed:
Merge branch 'sf/deepseek-ocr' of github.com:sfallah/llama.cpp into sf/deepseek-ocr

bluebread commented on tools/mtmd/clip.cpp:
```diff
@@ -4423,6 +5275,182 @@ bool clip_image_preprocess(struct clip_ctx * ctx, const clip_image_u8 * img, str
+                    res_imgs->grid_y = 1;
+                }
+                else {
+                    GGML_ABORT("DeepSeek-OCR: Gundam/Gundam-Master haven't been tested yet.\n");
```
@ngxson I've encountered an issue with batching images. In order to handle images much larger than 1280x1280, DeepSeek-OCR crops them into 640x640 (gundam) or 1024x1024 (gundam master) subimages as local views. However, the current framework doesn't support batching multiple images. Technically, it shouldn't be too difficult to add batch support, but I'm concerned about introducing new bugs and affecting other models. Do you have any suggestions?

ngxson replied:
IIRC it should be the same logic as llava uhd or minicpm-v where imae is cropped into smaller sub-images

batching is not yet supported, but does all sub-images need to be on the same batch?

otherwise, what we can do is to extend the `clip_image_f32` to include a notion of "nz":

```
struct clip_image_f32 {
    int nx;
    int ny;
    int nz; // can be > 1 for deepseek ocr

    std::vector<float> buf;
};
```

And memory layout corresponding to what you need on cgraph (to avoid another `ggml_permute` for example)

bluebread committed:
mtmd: remove --dsocr-mode argument

bluebread committed:
mtmd: refactor code & remove unused helper functions

sfallah commented:
FYI:
Additional quant versions of the models are available on HF [sabafallah/DeepSeek-OCR-GGUF](https://huggingface.co/sabafallah/DeepSeek-OCR-GGUF).

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -5990,6 +6001,99 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
+                self.preprocessor_config = json.load(f)
+
+
+    def set_gguf_parameters(self):
```
```suggestion
    def set_gguf_parameters(self):
```
No longer required, already handled.

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -7219,31 +7324,49 @@ def set_vocab(self):
+        kv_lora_rank = hparams["q_lora_rank"] if hparams["q_lora_rank"] is not None else 512
+        routed_scaling_factor = hparams.get("routed_scaling_factor", 1.0)
+        norm_topk_prob = hparams.get("norm_topk_prob", False)
+        scoring_func = hparams.get("scoring_func", "softmax")
```
```suggestion
```
Already handled.

ngxson commented on gguf-py/gguf/constants.py:
```diff
@@ -1057,6 +1077,22 @@ class MODEL_TENSOR(IntEnum):
+    MODEL_TENSOR.V_SAM_NET_2:               "v.sam.net_2",
+    MODEL_TENSOR.V_SAM_NET_3:               "v.sam.net_3",
+    MODEL_TENSOR.V_ENC_EMBD_IMGNL:          "model.image_newline", # Deepseek-OCR
+    MODEL_TENSOR.V_ENC_EMBD_VSEP:           "model.view_seperator", # Deepseek-OCR
```
please prefix any vision-related tensor with `v.`

```
v.image_newline
v.view_sep
```

CISC commented on examples/eval-callback/eval-callback.cpp:
Revert changes.

CISC commented on gguf-py/gguf/tensor_mapping.py:
```diff
@@ -2,6 +2,8 @@
 
+from numpy.f2py.auxfuncs import throw_error
+
 from .constants import MODEL_ARCH, MODEL_TENSOR, MODEL_TENSORS, TENSOR_NAMES
```
```suggestion
from .constants import MODEL_ARCH, MODEL_TENSOR, MODEL_TENSORS, TENSOR_NAMES
```

ngxson commented on tools/mtmd/clip.cpp:
```diff
@@ -3900,6 +4385,209 @@ struct img_tool {
 
+    // Bicubic resize function using Pillow's ImagingResample algorithm
+    // Adapted from https://github.com/python-pillow/Pillow/blob/main/src/libImaging/Resample.c
+    static bool resize_bicubic_pillow(const clip_image_u8 & img, clip_image_u8 & dst, int target_width, int target_height) {
```
please include some comments to explain the algorithmic/numeric differences between `RESIZE_ALGO_BICUBIC_PILLOW` and `RESIZE_ALGO_BICUBIC`

ngxson reviewed (changes requested):
I don't have time to fully review this PR yet, but I think it still need some clean up. Will have a deeper look next week.

In the meantime, would appreciate if you can try to simplify some parts of the code. Skim through the PR and remove any redundant code or comments

sfallah commented on tools/mtmd/clip.cpp:
```diff
@@ -1223,7 +1302,7 @@ struct clip_graph {
                                 hparams.ffn_op,
                                 model.position_embeddings,
-                                nullptr);
+                                nullptr); // shape [1024, 16, 16]
```
@ngxson
why, to the comment, or that add_ops is NULL?
If it is about the comments with "shape ....".
I tried to track the shapes, but not only I remove all those, they are also mostly wrong.

bluebread committed:
mtmd: fix tensor names for image newlines and view separator

ngxson commented:
@sfallah This case cannot be considered as trivial auto-completion. An auto-complete only counts if the AI generates things that you already had in mind - it just helps you skip the step to type from your mind to the keyboard. If the function is originally written by AI, that means you don't yet have the exact code in mind at that point.

Since the original function is written by chatgpt, there can be instances that it starts with something looks good but potentially be wrong (and your work built on top of that, which can also inherit these faults)

So, please explicitly which function is originally generated by AI

sfallah committed:
clean up

sfallah committed:
Merge branch 'sf/deepseek-ocr' into sf/deepseek-ocr-cleanup

# Conflicts:
#	gguf-py/gguf/tensor_mapping.py

ngxson commented:
Probably don't need for now, I just need an explicit mention because it will make reviewing easier

sfallah committed:
reverting automatically removed spaces

sfallah committed:
reverting automatically removed spaces

bluebread committed:
mtmd: fixed bad ocr check in Deepseek2 (LM)

bluebread committed:
Merge branch 'sf/deepseek-ocr-cleanup' of github.com:sfallah/llama.cpp into sf/deepseek-ocr-cleanup

bluebread committed:
mtmd: support combined QKV projection in buid_vit

sfallah committed:
Merge pull request #8 from sfallah/sf/deepseek-ocr-cleanup

Sf/deepseek ocr cleanup

sfallah committed:
using common build_attn in sam

sfallah committed:
corrected code-branch when flash-attn disabled
enabling usage of --flash-attn option

bluebread committed:
mtmd: minor fix

sfallah committed:
minor formatting and style

sfallah committed:
Merge pull request #9 from sfallah/sf/deepseek-ocr-attn

using common build_attn in sam

CISC commented on tools/mtmd/clip.cpp:
```diff
@@ -2073,19 +2152,44 @@ struct clip_graph {
+
+                    Qcur = ggml_cont(ctx0, ggml_view_3d(ctx0, QKV, ne0, ne2, ne3, nb2, nb3, 0*nb1));
+                    Kcur = ggml_cont(ctx0, ggml_view_3d(ctx0, QKV, ne0, ne2, ne3, nb2, nb3, 1*nb1));
+                    Vcur = ggml_cont(ctx0, ggml_view_3d(ctx0, QKV, ne0, ne2, ne3, nb2, nb3, 2*nb1));
```
Avoid `cont`, since there's no Q/K norm the easiest way is to move Q/K norm and reshape into the `else`.

sfallah committed:
Merge branch 'ggml-org:master' into sf/deepseek-ocr

sfallah committed:
fixed flake8 lint issues

sfallah committed:
minor editorconfig-check fixes

sfallah committed:
minor editorconfig-check fixes

bluebread reviewed (commented):
I'll postpone gundam mode support to a separate PR (probably with llama-server) to make this PR easier to review, since we have gotten base/large modes, which are good enough to handle most of the use cases.

bluebread commented on tools/mtmd/mtmd-cli.cpp:
```diff
@@ -313,8 +319,18 @@ int main(int argc, char ** argv) {
+                    image_tokens += '\n';
+                }
+                params.prompt = image_tokens + params.prompt;
+            }
```
I added this because DeepSeek-OCR wouldn't work if image tokens follow text tokens (see https://github.com/sfallah/llama.cpp/pull/5), but this doesn't look like a elegant approach to fix it. I'd like to hear you guys thoughts on this.

bluebread commented on tools/mtmd/clip.cpp:
```diff
@@ -2173,6 +2277,150 @@ struct clip_graph {
+        x = ggml_cont(m, x);
+        return x;
+    }
+
```
I extended ggml_get_rel_pos, ggml_win_part, ggml_win_unpart (which are not new ops and have already been there) in PR #17383, but now that seems unnecessary. Should we move this code to GGML, or handle it differently?

ngxson replied:
Just a quick look, I think the construction of `idx_2d` is better done on CPU as an input node instead. The translation (using AI) from pytorch to ggml looks overkill, while providing sub-optimal performance due to many ops not supported by backend, as well as many redundant `ggml_cont` and `ggml_repeat`

IMO, you just need to create one single absolute coordinates `[q_size, k_size]` and shift them by using `ggml_scale_bias`

Re. `window_partition`, can this be a simple attention mask like what Qwen2.5 uses instead?

bluebread replied:
Yes, `window_partition` can be rewritten as mask, but wouldn't it hurt performance? If we don't reshape QKV into window batches, `build_attn` adds the attn mask to QK as a bias and still has to compute the full matmul. 
Thanks for the suggestion on `get_rel_pos`. It makes a lot of sense. I'll work on that later.

sfallah replied:
@ngxson
Let me first say that I respect your expertise a great deal.
And I am a big fan of your great contributions to *llama.cpp*, but especially *mtmd*.
I am also grateful for being part of this PR to learn from all of you.

But if you allow me, I would like to ask two genuine questions:

1. How much of an improvement would it be to rewrite the window-partitioning using a window-mask? Both in terms of performance and in terms of code clarity?
2. With the current state of the PR, how far are we from the finish line?

IMO we have the model logically/functionally implemented, with the exception of fully supporting batching. Based on my understanding of the architecture, the reference implementation, and the other implementations (`mlx-vlm`, `vLLM`, the new PR in `transformers`), I believe we have fulfilled the main feature request.
And again based on my understanding, the performance will definitely not be worse than any of the other implementations.

My humble suggestion is that, if you and the other maintainers don’t see any major issues left, we should consider merging this PR and focus on improvements in subsequent PRs.

ngxson replied:
> Performance may not change very much

Looking at the comment from @bluebread below, I suspect that some 4D tensors are not supported by some ops. So yes, probably it does take a big hit in performance since some ops are not supported by GPU.

bluebread committed:
mtmd: simplify get_rel_pos

bluebread committed:
mtmd: make sam hparams configurable

bluebread committed:
mtmd: add detailed comments for resize_bicubic_pillow

bluebread committed:
mtmd: fixed wrong input setting

bluebread committed:
mtmd: convert model in FP16

bluebread committed:
mtmd: minor fix

bluebread commented on tools/mtmd/mtmd-cli.cpp:
```diff
@@ -333,6 +349,11 @@ int main(int argc, char ** argv) {
         }
 
     } else {
+        if (mtmd_is_deepseekocr(ctx.ctx_vision.get())) {
```
Got it. Would you like me to clean it up or would you do it yourself?

ngxson commented:
I think we must make sure it works with different input image qualities (e.g. lighting conditions, colors) and different sizes. A test script would be nice-to-have - ~~it can be push to a https://gist.github.com/ and shared here in the comment.~~ such test script can be placed into `tools/mtmd/test-deepseek-ocr.py` for example

Any bugs related to output quality should be addressed before I can do any refactoring on the PR, otherwise it's very difficult to trace back the commit containing the bug.

bluebread committed:
mtmd: remove tweak to llama-mtmd-cli & deepseek-ocr template

sfallah committed:
fix: test-1.jpg ORC issue with small (640) resolution
setting min-resolution base (1024) max large (1280) for dynamic-resolution

sfallah committed:
minor: editconfig-check fix

sfallah committed:
Merge branch 'master' into sf/deepseek-ocr-merge-test

# Conflicts:
#	tools/mtmd/clip.cpp
#	tools/mtmd/mtmd-cli.cpp

sfallah committed:
merge with changes from https://github.com/ggml-org/llama.cpp/pull/17909
added new opt to tests.sh to disable flash-attn

sfallah committed:
minor: editconfig-check fix

sfallah committed:
testing deepseek-ocr
quick and dirty test script comparing results of Qwen2.5-VL vs DeepSeek-OCR

ngxson commented:
heads up, sorry for the breaking change but there will be a refactoring (just moving stuff around) in https://github.com/ggml-org/llama.cpp/pull/17965

after finishing with this refactoring (and after you done testing on your side), I'll go back to deepseek-ocr

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr-merge_#17965

# Conflicts:
#	src/llama-kv-cache.cpp
#	tools/mtmd/clip.cpp

sfallah committed:
quick and (potential) dirty merge with https://github.com/ggml-org/llama.cpp/pull/17909

sfallah committed:
refactoring, one single builder function and static helpers

sfallah committed:
added deepseek-ocr test to tests.sh

sfallah committed:
Merge pull request #11 from sfallah/sf/deepseek-ocr-merge_#17965

Merged with PR #17965

sfallah committed:
minor formatting fixes

sfallah committed:
check with fixed expected resutls

sfallah committed:
Merge pull request #10 from sfallah/sf/deepseek-ocr-test-script

python test script for deepseek-ocr
testing OCR on text-1.jpeg newspaper image
checking against expected reference model output for Free-OCR and Markdown

sfallah committed:
minor formatting

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	convert_hf_to_gguf.py

sfallah committed:
editorconfig-check fix

sfallah committed:
Merge branch 'ggml-org:master' into sf/deepseek-ocr

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	gguf-py/gguf/constants.py
#	gguf-py/gguf/tensor_mapping.py
#	tools/mtmd/clip-impl.h
#	tools/mtmd/clip.cpp
#	tools/mtmd/models/models.h

sfallah committed:
merge with changes from https://github.com/ggml-org/llama.cpp/pull/18042

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	src/llama-arch.cpp

sfallah committed:
minor
- added GLM-4.6V to big tests
- added missing deps for python test

bluebread committed:
convert: minor fix

bluebread commented on convert_hf_to_gguf.py:
```diff
@@ -7122,6 +7187,15 @@ def prepare_tensors(self):
+        if 'clip-l-14-224' in vision_config and 'sam_vit_b' in vision_config:
+            self.model_arch = gguf.MODEL_ARCH.DEEPSEEK2OCR
+            self.gguf_writer.arch = gguf.MODEL_ARCH_NAMES[self.model_arch]
+            self.gguf_writer.add_architecture()
```
Fixed it. Thanks!

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -7177,44 +7252,62 @@ def set_vocab(self):
-        # note: deepseek2 using MLA converts into MQA (ie: GQA with 1 group)
-        self.hparams["num_key_value_heads"] = 1
+        if (rope_mscale_all := self.rope_parameters.get("mscale_all_dim")) is not None:
+            self.hparams["num_key_value_heads"] = self.hparams.get("rms_norm_eps", 1e-6)
```
This does not look right. :)

bluebread replied:
omg totally rushed that. This time should be correct (facepalm)...

CISC replied:
Uhhh, not so sure, still seems off. :)

CISC replied:
I was just wondering why you were checking (and assigning) `mscale_all_dim`, then not using it but instead presumably setting a default `rms_norm_eps`. If all you wanted to do was the latter, all is good now.

bluebread committed:
mtmd: format code

bluebread committed:
convert: quick fix

bluebread committed:
convert: quick fix

sfallah committed:
minor python formatting

bluebread commented:
@ngxson @CISC 
After a few days of investigation, I found that there might be no easy way to add support for DeepSeek-OCR Gundam modes. Probably I'm just overthinking, so I'd be happy to hear any suggestions on this. In my understanding, the whole image processing in gundam modes is as follows:

1. Resize image to height & width of multiples of 640 and then crop into 640x640 sub-images as local views. 
2. Resize & pad image to 1024x1024 as global view. 
3. Map all images (including sub-images) to vision tokens and concatenate them directly (without any other tokens between). 
4. Append view separator embedding (which is not a special token in the vocabulary). 

So here are two options that I can come up with:

1. **Add view separator in the vocabulary**: During gguf conversion (convert_hf_to_gguf.py), we add a new special token <view_separator> and insert view separator tensor in the embeddings table. Then, apply llava-uhd style processing to map global & local views into vision tokens one at a time, similar to llama4 or minicpmv. 
2. **Batch local views as additional 3D input and do all the stuff in the model**: Slightly extend clip_image_f32 by adding `nz` field (as @ngxson suggested before) and store local views in a single batch. Then **somehow** pass the batch of local views to the clip_image_batch_encode function and set it as inputs to the model. Finally handle everything (including view separator) in the model. 

IMO both options have their own pros & cons:

1. The first approach might be simpler to implement, but I am not sure if it's really appropriate to modify the embeddings table and the vocabulary. Moreover, I don't know what would happen if the vocabulary size becomes larger than LM originally expected. 
2. The second way should be more efficient (due to batching local views) and easier to debug (because it mirrors the implementation of the reference model). However, there seems to be no **elegant** way to pass local views from clip_image_preprocess to clip_image_batch_encode without modifying the framework.

sfallah committed:
Merge branch 'master' into sf/deepseek-ocr

sfallah committed:
fixed merge build issue

ngxson commented:
> we add a new special token <view_separator> and insert view separator tensor in the embeddings table.

In case the token is not in vocab, we can just append its embedding as part of the image embedding itself.

You can modify image_f32 object to include a "is_add_termination" boolean. If set to yes, the output embedding will be concat with the token embedding. The same logic with boi/eoi tokens on glm-edge

sfallah commented:
> > > Got it. I'll work on the second approach. Thanks!
> > 
> > 
> > 
> > @bluebread 
> > 
> > FYI: I had just started with it yesterday:
> > 
> > https://github.com/sfallah/llama.cpp/tree/sf/deepseek-ocr-gundam-batch
> > 
> > It is still in progress!
> 
> @sfallah 
> Great job! I've also been working on this and implemented the modes but haven't 
> published yet (only missing the view separator token handling). I'll push my 
> branch later. Would you like to coordinate or compare it with your branch? Happy to help 
> if useful!
> 
@bluebread 

Great💪
I will not continue my branch and will wait for you to finish.

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	convert_hf_to_gguf.py
#	gguf-py/gguf/gguf_writer.py
#	gguf-py/gguf/tensor_mapping.py
#	src/llama-model.cpp
#	src/models/deepseek2.cpp
#	tools/mtmd/CMakeLists.txt
#	tools/mtmd/clip-impl.h
#	tools/mtmd/clip.cpp
#	tools/mtmd/clip.h

sfallah committed:
merge resolved
- fixed issues in convert
- tested several deepseek models

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

sfallah commented:
@aviallon 
It is not dead.
I am working on it.

sfallah committed:
minor fix

sfallah committed:
Merge branch 'ggml-org:master' into sf/deepseek-ocr

sfallah committed:
Merge branch 'ggml-org:master' into sf/deepseek-ocr

aviallon commented:
@sfallah is your branch ready?

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	tools/mtmd/clip.cpp

sfallah committed:
minor

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	convert_hf_to_gguf.py
#	gguf-py/gguf/tensor_mapping.py

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	convert_hf_to_gguf.py
#	gguf-py/gguf/tensor_mapping.py
#	src/llama-model.cpp

sfallah committed:
Update convert_hf_to_gguf.py

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

CISC reviewed (approved):
@ngxson has final approval

ngxson commented on tools/mtmd/clip-impl.h:
```diff
@@ -515,6 +534,32 @@ static std::string gguf_kv_to_str(const struct gguf_context * ctx_gguf, int i) {
+        t->name, ggml_type_name(t->type), ggml_op_desc(t),
+        src0->name, to_ne_string(src0).c_str(),
+        src1 ? src1_str : "");
+}
```
we should already had `string_format()` for this

ngxson commented on tools/mtmd/clip.cpp:
```diff
@@ -3266,6 +3623,89 @@ bool clip_image_preprocess(struct clip_ctx * ctx, const clip_image_u8 * img, str
+                // Use mean color for padding
+                unsigned char pad_r = static_cast<unsigned char>(params.image_mean[0] * 255.0f);
+                unsigned char pad_g = static_cast<unsigned char>(params.image_mean[1] * 255.0f);
+                unsigned char pad_b = static_cast<unsigned char>(params.image_mean[2] * 255.0f);
```
`img_tool::resize` already support custom padding color, why this is needed?

ngxson reviewed (commented):
I feel like many things can be simplified further. This PR still have too much code for a single-model support, that's what make me quite worry about the long-term direction of `mtmd`.

sfallah commented on tools/mtmd/clip.h:
```diff
@@ -108,8 +108,11 @@ bool clip_image_batch_encode(struct clip_ctx * ctx, int n_threads, const struct
 // note for contributor: this clip_is_(model) pattern is deprecated
 //                       do NOT add new functions like this
+bool clip_is_deepseekocr(const struct clip_ctx * ctx);
 
```
Sorry, I thought I had a few excuses.
I’ll handle it right, now.
Do you want me to remove the other `_clip_is__` functions while I’m at it?

ngxson replied:
No. You should only include changes related to deepseek ocr in this PR.

This PR already contains some non-related changes, it makes reviewing more complicated. Please extract any non-related changes to dedicated PR, otherwise I cannot merge this as-is

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	tools/mtmd/clip.cpp

sfallah committed:
- removed clip_is_deepseekocr
- removed redundant RESIZE_ALGO_BICUBIC_PILLOW resize-algo
- simplified image-preprocessing
- removed/simplified debug functions

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

sfallah committed:
- cleaning commented out code

sambartik commented:
Is this intended to work with `llama-server`? The command I used: 
```
./build/bin/llama-server -hf sabafallah/DeepSeek-OCR-GGUF --chat-template deepseek-ocr --temp 0 -fit on
```

As of `e037b95628c579265768fba12b103c51ac3bb446` it fails to accept a prompt

> <|grounding|>Convert the document to markdown.

with an attached test jpeg image. I'm seeing this in the server logs:

```
main: model loaded
main: server is listening on http://127.0.0.1:8080
main: starting the main loop...
srv  update_slots: all slots are idle
tokenize: error: number of bitmaps (1) does not match number of markers (0)
srv  log_server_r: done request: POST /v1/chat/completions 127.0.0.1 400
```

<img width="1250" height="931" alt="image" src="https://github.com/user-attachments/assets/83a2eda8-24af-4f2c-a638-724fbb03907b" />

Looks like that might be because of the chat template?

Also, I am getting this in the logs too. Not sure if it's connected to this PR, but I have not seen this in other models with master branch version of llama.cpp.

```
load_hparams: model size:         774.24 MiB
load_hparams: metadata size:      0.17 MiB
warmup: warmup with image size = 1024 x 1024
alloc_compute_meta:      CUDA0 compute buffer size =  1222.39 MiB
alloc_compute_meta:        CPU compute buffer size =   420.02 MiB
alloc_compute_meta: graph splits = 25, nodes = 1516
warmup: *****************************************************************
warmup: WARNING: flash attention not supported by CUDA0, memory usage will increase
warmup: op params: 
warmup:    dst: type = f32, ne = [64 12 4096 1], nb = [4 256 3072 12582912]
warmup:   src0: type = f32, ne = [64 4096 12 1], nb = [4 3072 256 12582912]
warmup:   src1: type = f16, ne = [64 4096 12 1], nb = [2 128 524288 6291456]
warmup:   src2: type = f16, ne = [64 4096 12 1], nb = [2 128 524288 6291456]
warmup: please report this on github as an issue
warmup: *****************************************************************
alloc_compute_meta:      CUDA0 compute buffer size =  1982.14 MiB
alloc_compute_meta:        CPU compute buffer size =   420.02 MiB
alloc_compute_meta: graph splits = 1, nodes = 1588
warmup: flash attention is disabled
warmup: *****************************************************************
warmup: WARNING: the CLIP graph uses unsupported operators by the backend
warmup:          the performance will be suboptimal                      
warmup:          list of unsupported ops (backend=CUDA0):
warmup:         SOFT_MAX: type = f32, ne = [196 196 12 25]
warmup:             CONT: type = f32, ne = [196 64 12 25]
warmup:          PERMUTE: type = f32, ne = [64 4096 12 1]
warmup:              CPY: type = f16, ne = [196 196 12 25]
warmup:              ADD: type = f32, ne = [14 14 196 300]
warmup:          MUL_MAT: type = f32, ne = [64 64 64 12]
warmup:          RESHAPE: type = f32, ne = [14 1 196 300]
warmup:          PERMUTE: type = f32, ne = [14 14 14 300]
warmup:             CONT: type = f32, ne = [64 64 64 12]
warmup:          RESHAPE: type = f32, ne = [64 14 14 300]
warmup:          PERMUTE: type = f32, ne = [64 196 12 25]
warmup:         GET_ROWS: type = f32, ne = [64 4096 1 1]
warmup: flash attention is disabled
warmup: please report this on github as an issue
warmup: ref: https://github.com/ggml-org/llama.cpp/pull/16837#issuecomment-3461676118
warmup: *****************************************************************

```

Running it with `llama-mtmd-cli` works, though.

sambartik commented:
Have you tried with `--chat-template deepseek-ocr` as you suggested in the GGUF Readme and this PR description @sfallah? Using the default chat template results in looping issues.

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	src/llama-model.cpp

sfallah committed:
fixing instabilities issues reintroducing resize_bicubic_pillow

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	src/llama-model.cpp

sfallah committed:
- use f16 model for deepseek-ocr test
- ignore llama-arch test for deepseek-ocr

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	tools/mtmd/clip.cpp

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	convert_hf_to_gguf.py
#	src/llama-model.cpp

sfallah committed:
Merge remote-tracking branch 'sfallah/master' into sf/deepseek-ocr

# Conflicts:
#	tools/mtmd/models/glm4v.cpp
#	tools/mtmd/models/siglip.cpp

ngxson committed:
rename fc_w --> mm_fc_w

ngxson committed:
Merge branch 'master' into sf/deepseek-ocr

ngxson committed:
add links to OCR discussion

ngxson committed:
cleaner loading code

ngxson committed:
add missing .weight to some tensors

ngxson committed:
add default jinja template (to be used by server)

ngxson committed:
move test model to ggml-org

sfallah committed:
rolling back upscale change

ngxson commented:
Test results look good, PTAL @CISC @ggerganov (I need 2 approvals)

```
[vision] OK:   ggml-org/SmolVLM-500M-Instruct-GGUF:Q8_0
[vision] OK:   ggml-org/SmolVLM2-2.2B-Instruct-GGUF:Q4_K_M
[vision] OK:   ggml-org/SmolVLM2-500M-Video-Instruct-GGUF:Q8_0
[vision] OK:   ggml-org/gemma-3-4b-it-GGUF:Q4_K_M
[vision] OK:   THUDM/glm-edge-v-5b-gguf:Q4_K_M
[vision] OK:   second-state/Llava-v1.5-7B-GGUF:Q2_K
[vision] OK:   cjpais/llava-1.6-mistral-7b-gguf:Q3_K_M
[vision] OK:   ibm-research/granite-vision-3.2-2b-GGUF:Q4_K_M
[vision] OK:   second-state/MiniCPM-Llama3-V-2_5-GGUF:Q2_K
[vision] OK:   openbmb/MiniCPM-V-2_6-gguf:Q2_K
[vision] OK:   openbmb/MiniCPM-o-2_6-gguf:Q4_0
[vision] OK:   bartowski/Qwen2-VL-2B-Instruct-GGUF:Q4_K_M
[vision] OK:   ggml-org/Qwen2.5-VL-3B-Instruct-GGUF:Q4_K_M
[vision] OK:   ggml-org/InternVL2_5-1B-GGUF:Q8_0
[vision] OK:   ggml-org/InternVL3-1B-Instruct-GGUF:Q8_0
[vision] OK:   ggml-org/Qwen2.5-Omni-3B-GGUF:Q4_K_M
[vision] OK:   ggml-org/LFM2-VL-450M-GGUF:Q8_0
[vision] OK:   ggml-org/granite-docling-258M-GGUF:Q8_0
[vision] OK:   ggml-org/LightOnOCR-1B-1025-GGUF:Q8_0
[vision] OK:   ggml-org/DeepSeek-OCR-GGUF:Q8_0
[audio]  OK:   ggml-org/ultravox-v0_5-llama-3_2-1b-GGUF:Q8_0
[audio]  OK:   ggml-org/Qwen2.5-Omni-3B-GGUF:Q4_K_M
[audio]  OK:   ggml-org/Voxtral-Mini-3B-2507-GGUF:Q4_K_M
[audio]  OK:   ggml-org/LFM2-Audio-1.5B-GGUF:Q8_0
```

ggerganov reviewed (approved):
`libllama` changes are OK

ngxson committed:
Update convert_hf_to_gguf.py

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

urumchi commented:
thanks autho's work , but i meet some problem ,please someone could help me ?
https://huggingface.co/sahilchachra/Unlimited-OCR-GGUF/discussions/4

the error is :
```
~/sources/llama.cpp | pr24975  ./build/bin/llama-mtmd-cli \                                                                                  ok | 5s | base py | 19:06:17
  -m /Users/apple/Documents/models/Unlimited-OCR-GGUF/Unlimited-OCR-Q4_K_M.gguf \
  --mmproj /Users/apple/Documents/models/Unlimited-OCR-GGUF/mmproj-Unlimited-OCR-F16.gguf \
  --image /Users/apple/Downloads/奥赛数学题.png \
  -p "<|grounding|>Convert the document to markdown." \
  --temp 0
0.00.062.324 I common_init_result: fitting params to device memory ...
0.00.062.327 I common_init_result: (for bugs during this step try to reproduce them with -fit off, or provide --verbose logs if the bug only occurs with -fit on)
0.00.413.335 I common_init_from_params: warming up the model with an empty run - please wait ... (--no-warmup to disable)
WARNING: Using native backtrace. Set GGML_BACKTRACE_LLDB for more info.
WARNING: GGML_BACKTRACE_LLDB may cause native MacOS Terminal.app to crash.
See: https://github.com/ggml-org/llama.cpp/pull/17869
0   libggml-base.0.15.2.dylib           0x00000001027d6e74 ggml_print_backtrace + 276
1   libggml-base.0.15.2.dylib           0x00000001027ecdc4 _ZL23ggml_uncaught_exceptionv + 12
2   libc++abi.dylib                     0x0000000186310698 _ZSt11__terminatePFvvE + 16
3   libc++abi.dylib                     0x0000000186313c30 __cxa_get_exception_ptr + 0
4   libc++abi.dylib                     0x0000000186313bd8 _ZN10__cxxabiv1L12failed_throwEPNS_15__cxa_exceptionE + 0
5   libllama-common.0.0.9771.dylib      0x0000000102ebcaec _Z27common_chat_templates_applyPK21common_chat_templatesRK28common_chat_templates_inputs + 2172
6   libllama-common.0.0.9771.dylib      0x0000000102ec1c84 _Z26common_chat_format_examplePK21common_chat_templatesbRKNSt3__13mapINS2_12basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEES9_NS2_4lessIS9_EENS7_INS2_4pairIKS9_S9_EEEEEE + 648
7   llama-mtmd-cli                      0x00000001025c863c _ZN16mtmd_cli_contextC2ER13common_params + 516
8   llama-mtmd-cli                      0x00000001025c5024 main + 208
9   dyld                                0x0000000185fbab98 start + 6076
libc++abi: terminating due to uncaught exception of type std::runtime_error: this custom template is not supported, try using --jinja
[1]    9610 abort      ./build/bin/llama-mtmd-cli -m  --mmproj  --image  -p  --temp 0
```

sfallah commented:
@urumchi 
are you trying the PR https://github.com/ggml-org/llama.cpp/pull/24975?

Please do use the following GGUF model.

GGUF models: [sabafallah/Unlimited-OCR-GGUF](https://huggingface.co/sabafallah/Unlimited-OCR-GGUF)


```bash
build/bin/llama-mtmd-cli -hf sabafallah/Unlimited-OCR-GGUF:bf16 \
  --image tools/mtmd/test-1.jpeg -p "document parsing." \
  --chat-template deepseek-ocr \
  --temp 0 --flash-attn off --no-warmup \
  -n 4096 -c 16384 \
  --dry-multiplier 0.8 --dry-base 1.75 --dry-allowed-length 2 \
  --dry-penalty-last-n -1 --dry-sequence-breaker none
```

PS: the https://github.com/ggml-org/llama.cpp/pull/24975 is still a draft, but as of https://github.com/ggml-org/llama.cpp/pull/24969 you can you use the model with llama.cpp releases.
